pkg/operator/ceph/object/shared_pools.go

```diff
@@ -0,0 +1,612 @@
+	"reflect"
+	"sort"
+	"strings"
+
+	"github.com/pkg/errors"
+	cephv1 "github.com/rook/rook/pkg/apis/ceph.rook.io/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+)
+
+const (
+	defaultPlacementName           = "default"
+	defaultPlacementCephConfigName = "default-placement"
+	defaultPlacementStorageClass   = "STANDARD"
+)
+
+func IsNeedToCreateObjectStorePools(sharedPools cephv1.ObjectSharedPoolsSpec) bool {
+	for _, pp := range sharedPools.PoolPlacements {
+		if pp.Name == defaultPlacementName {
+			// No need to create pools. External pools from default placement will be used
+			return false
+		}
+	}
+	if sharedPools.MetadataPoolName != "" && sharedPools.DataPoolName != "" {
+		// No need to create pools. Shared pools will be used
+		return false
+	}
+	return true
+}
+
+func validatePoolPlacements(placements []cephv1.PoolPlacementSpec) error {
+	names := make(map[string]struct{}, len(placements))
+	for _, p := range placements {
+		if _, ok := names[p.Name]; ok {
+			return fmt.Errorf("invalidObjStorePoolCofig: invalid placement %s: placement names must be unique", p.Name)
+		}
+		names[p.Name] = struct{}{}
+		if err := validatePoolPlacementStorageClasses(p.StorageClasses); err != nil {
+			return fmt.Errorf("invalidObjStorePoolCofig: invalid placement %s: %w", p.Name, err)
+		}
+	}
+	return nil
+}
+
+func validatePoolPlacementStorageClasses(scList []cephv1.PlacementStorageClassSpec) error {
+	names := make(map[string]struct{}, len(scList))
+	for _, sc := range scList {
+		if sc.Name == defaultPlacementStorageClass {
+			return fmt.Errorf("invalid placement StorageClass %q: %q name is reserved", sc.Name, defaultPlacementStorageClass)
+		}
+		if _, ok := names[sc.Name]; ok {
+			return fmt.Errorf("invalid placement StorageClass %q: name must be unique", sc.Name)
+		}
+		names[sc.Name] = struct{}{}
+	}
+	return nil
+}
+
+func adjustZonePlacementPools(zone map[string]interface{}, spec cephv1.ObjectSharedPoolsSpec) (hasChanged bool, err error) {
+	name, err := getObjProperty[string](zone, "name")
+	if err != nil {
+		return false, fmt.Errorf("unable to get zone name: %w", err)
+	}
+
+	placements, err := getObjProperty[[]interface{}](zone, "placement_pools")
+	if err != nil {
+		return false, fmt.Errorf("unable to get pool placements for zone %s: %w", name, err)
+	}
+	fromSpec := toZonePlacementPools(spec, name)
+
+	inConfig := map[string]struct{}{}
+	idxToRemove := map[int]struct{}{}
+	for i, p := range placements {
+		pObj, ok := p.(map[string]interface{})
+		if !ok {
+			return false, fmt.Errorf("unable to cast pool placement to object for zone %s: %+v", name, p)
+		}
+		placementID, err := getObjProperty[string](pObj, "key")
+		if err != nil {
+			return false, fmt.Errorf("unable to get pool placement name for zone %s: %w", name, err)
+		}
+		// check if placement should be removed
+		if _, inSpec := fromSpec[placementID]; !inSpec && placementID != defaultPlacementCephConfigName {
+			// remove non-default placement if it is not in spec
```

Comment:
Nit: I got tripped up understanding this helpful comment. Suggesting slightly clearer wording.

```suggestion
			// remove placement if it is not in spec, but don't remove default placement
```